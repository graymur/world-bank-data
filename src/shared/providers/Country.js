import React from 'react';
import PropTypes from 'prop-types';

import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

export const countryQuery = gql`
    query countryQuery($iso2Code: String!) {
        country(iso2Code: $iso2Code) {
            name
            iso2Code
            region
            incomeLevel
            capitalCity
        }
    }
`;

export class CountryProvider extends React.Component {
	static propTypes = {
		data: PropTypes.object,
		render: PropTypes.func
	};

	render() {
		const {loading, country} = this.props.data;
		return this.props.render({loading, country});
	}
}

export default graphql(countryQuery, {
	options: ownProps => ({ variables: { iso2Code: ownProps.iso2Code } })
})(CountryProvider);
