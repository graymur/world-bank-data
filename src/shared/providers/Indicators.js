import React from 'react';
import PropTypes from 'prop-types';

import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

const indicatorsQuery = gql`
    query indicatorsQuery{
        indicators {
            id,
            name
        }
    }`;

export class IndicatorsProvider extends React.Component {
	static propTypes = {
		data: PropTypes.object,
		render: PropTypes.func
	};

	render() {
		const {loading, indicators} = this.props.data;
		return this.props.render({loading, indicators});
	}
}

export default graphql(indicatorsQuery)(IndicatorsProvider);
