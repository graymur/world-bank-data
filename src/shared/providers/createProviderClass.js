import React from 'react';
import PropTypes from 'prop-types';

export default dataFieldName => class extends React.Component {
	static propTypes = {
		data: PropTypes.object,
		render: PropTypes.func
	};

	render() {
		return this.props.render({
			loading: this.props.data.loading,
			[dataFieldName]: this.props.data[dataFieldName]
		});
	}
};
