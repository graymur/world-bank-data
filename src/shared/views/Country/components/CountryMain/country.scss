@import "~style/lib";

.country {
    &._loading {
        min-height: 20rem;
    }
}