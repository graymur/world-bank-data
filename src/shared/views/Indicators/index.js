import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {createStructuredSelector} from 'reselect';
import * as actions from 'shared/logic/indicators/actions';
import * as selectors from 'shared/logic/indicators/selectors';
import IndicatorsMain from './components/IndicatorsMain';
import IndicatorsProvider from 'shared/providers/Indicators';
import queryString from 'query-string';

export class Indicators extends React.Component {
	static propTypes = {
		location: PropTypes.object,
		setSearchString: PropTypes.func,
		history: PropTypes.shape({
			push: PropTypes.func.isRequired,
			location: PropTypes.object
		})
	};

	constructor(props) {
		super(props);
		this.initiateSearch = this.initiateSearch.bind(this);
		this.clearSearchResults = this.clearSearchResults.bind(this);
	}

	clearSearchResults() {
		this.props.history.push(`/indicators`);
	}

	initiateSearch(value) {
		this.props.history.push(`/indicators?search=${encodeURIComponent(value)}`);
	}

	componentDidMount() {
		const parsed = queryString.parse(this.props.location.search);

		if (parsed.search) {
			return this.props.setSearchString(parsed.search);
		}
	}

	render() {
		return <IndicatorsProvider render={data =>
			<IndicatorsMain {...this.props} {...data} clearSearchResults={this.clearSearchResults} initiateSearch={this.initiateSearch}/>
		}/>;
	}
}

const mapStateToProps = createStructuredSelector({
	searchString: selectors.searchString
});

export default connect(mapStateToProps, actions)(Indicators);
