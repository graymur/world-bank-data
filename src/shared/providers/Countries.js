import createProviderClass from './createProviderClass';

import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

const countriesQuery = gql`
	query { 
		countries { 
			name, 
			iso2Code
		}
	}
`;

export default graphql(countriesQuery)(createProviderClass('countries'));
