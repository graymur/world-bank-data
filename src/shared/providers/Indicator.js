import createProviderClass from './createProviderClass';

import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

export const indicatorQuery = gql`
    query indicatorQuery($id: String!) {
        indicator(id: $id) {
            id,
            name,
            sourceNote,
            sourceOrganization
        }
    }
`;

export default graphql(indicatorQuery, {
	options: ownProps => ({ variables: { id: ownProps.indicatorId } })
})(createProviderClass('indicator'));
