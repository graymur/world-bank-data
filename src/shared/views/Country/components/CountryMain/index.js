import React from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';
import Helmet from 'react-helmet';
import getPageTitle from 'shared/utils/getPageTitle';
import Loader from 'shared/components/Loader';
import IndicatorsProvider from 'shared/providers/Indicators';
import classnames from 'classnames';
import './country.scss';

export default class Country extends React.Component {
	static propTypes = {
		setTitle: PropTypes.bool,
		loading: PropTypes.bool,
		country: PropTypes.object
	};

	static defaultProps = {
		setTitle: true
	};

	renderCountry(country) {
		const {setTitle} = this.props;

		if (!country) {
			return null;
		}

		return (
			<div className='country'>
				{setTitle && <Helmet><title>{getPageTitle(country.name)}</title></Helmet>}
				<h1>{country.name}</h1>
				<p>Region: {country.region.value}</p>
				<p>Income level: {country.incomeLevel.value}</p>
				<p>Capital city: {country.capitalCity}</p>
				<h1>Indicators</h1>
				<nav className='indicators__list'>
					<IndicatorsProvider render={({indicators}) => (indicators || []).map(indicator => (
						<Link
							key={indicator.id} className='indicators__list__item'
							to={`/countries/${country.iso2Code}/indicator/${indicator.id}`}>{indicator.name}</Link>
					))}/>
				</nav>
			</div>
		);
	}

	render() {
		const {loading, country} = this.props;

		const classNames = classnames('country', {'loading': loading});

		return (
			<div className={classNames}>
				{loading ? <Loader/> : this.renderCountry(country)}
			</div>
		);
	}
}
