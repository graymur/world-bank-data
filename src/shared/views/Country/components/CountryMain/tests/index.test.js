/* eslint-disable */
import React from 'react';
import renderer from 'react-test-renderer';
import {StaticRouter} from 'react-router';
import {MockedProvider} from 'react-apollo/test-utils';
import CountryMain from '../index';
import {indicatorsQuery} from 'shared/providers/Indicators'
import {addTypenameToDocument} from 'apollo-utilities';

const props = {
	loading: false,
	country: {
		iso2Code: 'AA',
		name: 'Country name',
		region: {value: 'Region name'},
		incomeLevel: {value: 'Income level'},
		capitalCity: 'City name'
	}
};

const indicators = [{
	id: 1,
	name: 'Indicator 1'
}, {
	id: 2,
	name: 'Indicator 2'
}]

test('Renders country and indicators list', () => {
	const component = renderer.create(
		<MockedProvider mocks={[
			{
				request: {
					query: addTypenameToDocument(indicatorsQuery),
					variable: {cache: false}
				},
				result: {data: {indicators}}
			}
		]}>
			<StaticRouter location="someLocation" context={{}}>
				<CountryMain {...props}/>
			</StaticRouter>
		</MockedProvider>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

