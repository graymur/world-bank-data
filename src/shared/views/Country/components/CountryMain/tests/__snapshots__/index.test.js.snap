// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders country and indicators list 1`] = `
<div
  className="country"
>
  <div
    className="country"
  >
    <h1>
      Country name
    </h1>
    <p>
      Region: 
      Region name
    </p>
    <p>
      Income level: 
      Income level
    </p>
    <p>
      Capital city: 
      City name
    </p>
    <h1>
      Indicators
    </h1>
    <nav
      className="indicators__list"
    />
  </div>
</div>
`;
